File: src/0078.subsets.78/solution.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsets = function (nums) {
    function backTracking (list, nums, sequence, index) {
        sequence.push(list.slice());
        for (let i = index; i < nums.length; i++) {
            list.push(nums[i]);
            backTracking(list, nums, sequence, i + 1);
            list.pop();
        }
    }
    const sequence = [];
    backTracking([], nums, sequence, 0);
    return sequence;
};

File: src/0078.subsets.78/solution.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        backTracking(nums,0,new ArrayList<>(),result);
        return result;
    }
    
    private void backTracking(int[] nums,int index,List<Integer> accumulate, List<List<Integer>> result){
        if(index == nums.length){
            result.add(new ArrayList<>(accumulate));
            return;
        }
        backTracking(nums,index+1,accumulate,result);
        
        accumulate.addLast(nums[index]);
        backTracking(nums,index+1,accumulate,result);
        accumulate.removeLast();
    }
    
}